File: bench/diff_memcat_p.c
// SPDX-License-Identifier: GPL-2.0
// Tier-2.5 differential oracle: C original vs Rust translation, memcat_p.
// Reference extracted from lib/memcat_p.c (v7.1); kept byte-identical
// (malloc instead of kmalloc_array — allocator choice is not under
// test, the merge/reverse-fill logic is).
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static void **__memcat_p(void **a, void **b)
{
	void **p = a, **new;
	int nr;

	for (nr = 0, p = a; *p; nr++, p++)
		;
	for (p = b; *p; nr++, p++)
		;
	nr++;

	new = malloc(nr * sizeof(void *));
	if (!new)
		return NULL;

	for (nr--; nr >= 0; nr--, p = p == b ? &a[nr] : p - 1)
		new[nr] = *p;

	return new;
}

// Explicit LCG (same constants used across all bench/diff_*.c files).
static uint64_t lcg_state;
static uint32_t lcg_next(void)
{
	lcg_state = lcg_state * 6364136223846793005ULL + 1442695040888963407ULL;
	return (uint32_t)(lcg_state >> 32);
}

#define MAXLEN 20

int main(int argc, char **argv)
{
	long n = argc > 1 ? atol(argv[1]) : 5000;
	lcg_state = argc > 2 ? (uint64_t)atol(argv[2]) : 271828;

	for (long i = 0; i < n; i++) {
		int alen = lcg_next() % MAXLEN;
		int blen = lcg_next() % MAXLEN;

		// Use distinct, identifiable "pointer" values (as small ints
		// cast to void*, never dereferenced) so the merged order is
		// externally observable without needing real allocations for
		// the elements themselves.
		void *a[MAXLEN + 1];
		void *b[MAXLEN + 1];
		for (int k = 0; k < alen; k++)
			a[k] = (void *)(uintptr_t)(0x1000 + lcg_next() % 0xff0 + 1); // never 0
		a[alen] = NULL;
		for (int k = 0; k < blen; k++)
			b[k] = (void *)(uintptr_t)(0x2000 + lcg_next() % 0xff0 + 1);
		b[blen] = NULL;

		void **merged = __memcat_p(a, b);

		printf("memcat,%d,%d,%d,", alen, blen, merged != NULL);
		if (merged) {
			int idx = 0;
			while (merged[idx] != NULL) {
				printf("%s%lx", idx ? "-" : "", (unsigned long)(uintptr_t)merged[idx]);
				idx++;
			}
			printf(",%d\n", idx);
			free(merged);
		} else {
			printf(",-1\n");
		}
	}
	return 0;
}

File: bench/diff_memcat_p.rs
// SPDX-License-Identifier: GPL-2.0
//! Tier-2.5 differential oracle: Rust translation side, memcat_p.
//! Faithful copy of lib/memcat_p_rs.rs's algorithm (kernel-crate
//! bindings/export/kmalloc_array stripped — host Vec stands in for the
//! kernel allocation, merge/reverse-fill logic unchanged) — same
//! protocol/LCG as diff_memcat_p.c.

/// Faithful port of __memcat_p's counting + reverse-fill loop, over
/// index-based "pointers" (u64 values, 0 reserved as the NULL
/// terminator, matching the C's void* NULL-termination convention)
/// instead of raw pointers.
fn memcat_p(a: &[u64], b: &[u64]) -> Option<Vec<u64>> {
    // a and b are NUL-terminated (already, by construction of the caller
    // — trailing 0 present); find their lengths same as the C's counting
    // loops (nr_a elements before the 0, plus nr_b).
    let nr_a = a.iter().position(|&x| x == 0).unwrap();
    let nr_b = b.iter().position(|&x| x == 0).unwrap();
    let nr = nr_a + nr_b + 1; // +1 for the NULL terminator

    // kmalloc_array failure isn't modeled on the host (always succeeds);
    // matches every other bench/diff_*.rs's treatment of allocation.
    let mut new = vec![0u64; nr];

    // Reverse-fill: last slot is the NULL terminator (0), then walk
    // backwards through b (b[nr_b-1..0]), then through a (a[nr_a-1..0]),
    // exactly matching the C's `p = p == b ? &a[nr] : p - 1` pointer walk.
    new[nr - 1] = 0;
    for i in (0..nr_b).rev() {
        new[nr_a + i] = b[i];
    }
    for i in (0..nr_a).rev() {
        new[i] = a[i];
    }

    Some(new)
}

// Identical LCG to diff_memcat_p.c.
struct Lcg(u64);
impl Lcg {
    fn next(&mut self) -> u32 {
        self.0 = self
            .0
            .wrapping_mul(6364136223846793005)
            .wrapping_add(1442695040888963407);
        (self.0 >> 32) as u32
    }
}

const MAXLEN: usize = 20;

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let n: u64 = args.get(1).and_then(|a| a.parse().ok()).unwrap_or(5000);
    let seed: u64 = args.get(2).and_then(|a| a.parse().ok()).unwrap_or(271828);
    let mut rng = Lcg(seed);

    for _ in 0..n {
        let alen = (rng.next() as usize) % MAXLEN;
        let blen = (rng.next() as usize) % MAXLEN;

        let mut a = vec![0u64; alen + 1];
        let mut b = vec![0u64; blen + 1];
        for k in 0..alen {
            a[k] = 0x1000 + (rng.next() % 0xff0) as u64 + 1;
        }
        a[alen] = 0;
        for k in 0..blen {
            b[k] = 0x2000 + (rng.next() % 0xff0) as u64 + 1;
        }
        b[blen] = 0;

        let merged = memcat_p(&a, &b);

        print!("memcat,{},{},{},", alen, blen, merged.is_some() as i32);
        match merged {
            Some(m) => {
                let mut idx = 0usize;
                while m[idx] != 0 {
                    if idx != 0 {
                        print!("-");
                    }
                    print!("{:x}", m[idx]);
                    idx += 1;
                }
                println!(",{}", idx);
            }
            None => println!(",-1"),
        }
    }
}
